src/szk.rs

```diff
@@ -0,0 +1,565 @@
+            };
+            let (joint_rand_seed, joint_rand) =
+                self.derive_joint_rand(leader_joint_rand_part, helper_joint_rand_part);
+            (joint_rand, Some([joint_rand_seed, host_joint_rand_part]))
+        } else {
+            (Vec::new(), None)
+        };
+        let verifier_share = self.typ.query(
+            input_share,
+            flp_proof_share.as_ref(),
+            &query_rand,
+            &joint_rand,
+            2,
+        )?;
+        Ok(SzkQueryShare {
+            joint_rand_seed_and_part: jr_seed_and_part,
+            verifier: verifier_share,
+        })
+    }
+
+    /// Returns true if the verifier message indicates that the input from which
+    /// it was generated is valid.
+    fn decide(
+        &self,
+        verifier: &[T::Field],
+        l_jr_part: Option<Seed<SEED_SIZE>>,
+        h_jr_part: Option<Seed<SEED_SIZE>>,
+        joint_rand_seed_opt: Option<Seed<SEED_SIZE>>,
+    ) -> Result<bool, SzkError> {
+        // Check if underlying FLP proof validates
+        let check_flp_proof = self.typ.decide(verifier)?;
+        if !check_flp_proof {
+            return Ok(false);
+        }
+        // Check that joint randomness was properly derived from both
+        // aggregators' parts
+
+        match (joint_rand_seed_opt, l_jr_part, h_jr_part) {
+            (Some(joint_rand_seed), Some(leader_joint_rand_part), Some(helper_joint_rand_part)) => {
+                if joint_rand_seed
+                    != self.derive_joint_rand_seed(&leader_joint_rand_part, &helper_joint_rand_part)
+                {
+                    Ok(false)
+                } else {
+                    Ok(true)
+                }
+            }
+            (None, None, None) => Ok(true),
+            (_, _, _) => Ok(false),
+        }
+    }
+}
+
+mod tests {
+    use super::*;
+    use crate::field::Field128 as TestField;
+    use crate::field::FieldElementWithInteger;
+    use crate::flp::types::{Count, Sum};
+    use crate::flp::Type;
+    use rand::{thread_rng, Rng};
+
+    fn generic_szk_test<T: Type>(typ: T, meas: T::Measurement, valid: bool) {
+        let encoded_measurement = typ.encode_measurement(&meas).unwrap();
+        let mut nonce = [0u8; 16];
+        let mut verify_key = [0u8; 16];
+        let algorithm_id = 5;
+        let szk_typ = Szk::new_turboshake128(typ.clone(), algorithm_id);
+        thread_rng().fill(&mut verify_key[..]);
+        thread_rng().fill(&mut nonce[..]);
+        let prove_rand_seed = Seed::<16>::generate().unwrap();
+        let helper_seed = Seed::<16>::generate().unwrap();
+        let leader_seed_opt = if typ.joint_rand_len() > 0 {
+            Some(Seed::<16>::generate().unwrap())
+        } else {
+            None
+        };
+        let helper_input_share: Vec<T::Field> =
+            Prng::from_seed_stream(XofTurboShake128::seed_stream(&helper_seed, &[2u8; 8], &[]))
+                .take(szk_typ.typ.input_len())
+                .collect();
```

Comment:
We need to expand the random vector from a specific seed that can be passed to the aggregator, so I don't think we can use random_vector, which samples and then destoys its own seed.

Reply:
That's true in Prio3, and in Mastic the IDPF will do something similar, but in this test the source of the helper input share vector doesn't matter, since it's external to `Szk`. I noticed that this expansion uses `0202020202020202` as a domain separation tag, while, nothing else does.

Reply:
I see, thanks!